src/oidcendpoint/oidc/token_coop.py

```diff
@@ -196,14 +211,48 @@ def _refresh_token_post_parse_request(self, request, client_id="", **kwargs):
 
         return request
 
+    def _token_exchange_post_parse_request(self, request, client_id="", **kwargs):
+        """
+        This is where clients come for token exchange
+
+        :param request: The request
+        :param authn: Authentication info, comes from HTTP header
+        :returns:
+        """
+
+        request = TokenExchangeRequest(**request.to_dict())
+
+        # verify that the request message is correct
+        try:
+            request.verify(keyjar=self.endpoint_context.keyjar)
+        except (MissingRequiredAttribute, ValueError, MissingRequiredValue) as err:
+            return self.error_cls(error="invalid_request", error_description="%s" % err)
+
+        try:
+            keyjar = self.endpoint_context.keyjar
+        except AttributeError:
+            keyjar = ""
+
+        request.verify(keyjar=keyjar, opponent_id=client_id)
+
+        if "client_id" not in request:
+            request["client_id"] = client_id
+
+        logger.debug("%s: %s" % (request.__class__.__name__, sanitize(request)))
+
+        return request
+
     def _post_parse_request(self, request, client_id="", **kwargs):
-        if request["grant_type"] == "authorization_code":
+        grant_type = request["grant_type"]
+        self.check_grant_type_allowed(grant_type)
+        if grant_type == "authorization_code":
             return self._access_token_post_parse_request(request, client_id, **kwargs)
-        else:  # request["grant_type"] == "refresh_token":
-            if self.allow_refresh:
-                return self._refresh_token_post_parse_request(request, client_id, **kwargs)
-            else:
-                raise ProcessError("Refresh Token not allowed")
+        elif grant_type == "refresh_token":
+            return self._refresh_token_post_parse_request(request, client_id, **kwargs)
+        elif grant_type == "urn:ietf:params:oauth:grant-type:token-exchange":
+            return self._token_exchange_post_parse_request(request, client_id, **kwargs)
+        else:
+            raise ProcessError("Invalid grant type")
```

Comment:
Add the `grant_type` value to the exception?

Reply:
yes